storage: merge replica.evaluateProposal and replica.evaluateProposalInner

nvb opened the pull request:
This change includes a series of cleanup that eventually ends in the merging of `replica.evaluateProposal` and `replica.evaluateProposalInner`. Along the way, it addresses a number of TODOs that have been around since the beginning of proposer-evaluated KV.

I've made a number of attempts at this change, and each time I threw it away because it became difficult to convince myself that I wasn't clobbering any desired behavior. To address this, I've split the PR into a series of equivalent refactors. The one exception to this is that all errors now "fail fast", because #21140 removed the cases where a proposal would result in an error but still want to propose in order to lie down intents.

This is all a precursor to cleanly addressing https://github.com/cockroachdb/cockroach/issues/23942.

nvb committed:
storage: never return nil batch from evaluateWriteBatch

With the small change to evaluateWriteBatch, an assertion showed
that the returned batch is never nil.

Release note: None

nvb committed:
storage: remove wasWriting flag from evaluateProposalInner

An assertion showed that it was never needed.

Release note: None

nvb committed:
storage: simplify error handling in evaluateProposalInner

Proposals that return errors but still want to go through Raft to
lay down intents were refactored away in #21140. This allows us to
simplify the error handling in `evaluateProposalInner`.

Release note: None

nvb committed:
storage: merge evaluateProposal and evaluateProposalInner

This addresses a TODO and makes this code much easier to work
with.

Release note: None

nvb committed:
storage: remove WriteTooOld hack in TestStoreResolveWriteIntentPushOnRead

This addresses two TODOs.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/24250)
<!-- Reviewable:end -->

bdarnell commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 2 of 2 files at r5.
Review status: :shipit: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/24250#-:-L8cHapVr0xgkz2PuC2o:bnfp4nl)*
<!-- Sent from Reviewable.io -->

nvb committed:
storage: remove LocalResult.Err entirely

Previous refactoring was able to simplify proposal logic now that all
failed proposals are "fail-fast". This final commit removed the
`LocalResult.Err` field entirely because it is no longer needed.

In doing so, this refactor disentangles `txnWaitQueue.Enqueue` from the
`DontRetryPushTxnFailures` configuration.

Release note: None

nvanbenschoten commented:
Added one more commit to remove the `LocalResult.Err` field, now that it is never used. PTAL.

bdarnell commented:
LGTM

nvanbenschoten commented:
Thanks for the review @bdarnell.

tschottdorf commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> late to the party but thanks for doing this!

---

Reviewed 1 of 1 files at r1, 1 of 1 files at r2, 1 of 1 files at r3, 1 of 1 files at r4, 2 of 2 files at r5, 4 of 4 files at r6.
Review status: :shipit: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24250#-:-L8muuso715YY8wzPAOi:bxmzn2)*
<!-- Sent from Reviewable.io -->
